client-local/src/main/java/com/cloudera/livy/client/local/LocalClient.java

```diff
@@ -248,76 +251,60 @@ public void run() {
       String master = conf.get("spark.master");
+      LOG.info("Master : " + master);
+      launcher.setMaster(master);
       Preconditions.checkArgument(master != null, "spark.master is not defined.");
 
-      List<String> argv = Lists.newArrayList();
-
-      if (sparkHome != null) {
-        argv.add(new File(sparkHome, "bin/spark-submit").getAbsolutePath());
-      } else {
-        LOG.info("No spark.home provided, calling SparkSubmit directly.");
-        argv.add(new File(System.getProperty("java.home"), "bin/java").getAbsolutePath());
-
         if (master.startsWith("local") ||
               master.startsWith("mesos") ||
               master.endsWith("-client") ||
               master.startsWith("spark")) {
           String mem = conf.get("spark.driver.memory");
-          if (mem != null) {
-            argv.add("-Xms" + mem);
-            argv.add("-Xmx" + mem);
-          }
+          launcher.setConf("spark.driver.memory",mem);
 
           String cp = conf.get("spark.driver.extraClassPath");
           if (cp != null) {
-            argv.add("-classpath");
-            argv.add(cp);
+             launcher.setConf("spark.driver.extraClassPath",cp);
           }
 
           String libPath = conf.get("spark.driver.extraLibPath");
           if (libPath != null) {
-            argv.add("-Djava.library.path=" + libPath);
+             launcher.setConf("spark.driver.extraLibraryPath",libPath);
           }
 
           String extra = conf.get(DRIVER_OPTS_KEY);
           if (extra != null) {
             for (String opt : extra.split("[ ]")) {
               if (!opt.trim().isEmpty()) {
-                argv.add(opt.trim());
+                launcher.setConf("spark.driver.extraJavaOptions",opt.trim());
               }
             }
           }
         }
 
-        argv.add("org.apache.spark.deploy.SparkSubmit");
-      }
-
       if (master.equals("yarn-cluster")) {
         String executorCores = conf.get("spark.executor.cores");
         if (executorCores != null) {
-          argv.add("--executor-cores");
-          argv.add(executorCores);
+          launcher.setConf("spark.executor.cores",executorCores);
         }
 
         String executorMemory = conf.get("spark.executor.memory");
         if (executorMemory != null) {
-          argv.add("--executor-memory");
-          argv.add(executorMemory);
+          launcher.setConf("spark.executor.memory",executorMemory);
         }
 
         String numOfExecutors = conf.get("spark.executor.instances");
         if (numOfExecutors != null) {
-          argv.add("--num-executors");
-          argv.add(numOfExecutors);
+          launcher.setConf("spark.executor.instances",numOfExecutors);
         }
       }
 
-      argv.add("--properties-file");
-      argv.add(sparkConf.getAbsolutePath());
-      argv.add("--class");
-      argv.add(RemoteDriver.class.getName());
+      launcher.addAppArgs("--properties-file" + sparkConf.getAbsolutePath());
```

Comment:
My fault for telling otherwise earlier, but `sparkConf` contains Spark options and you should use `setPropertiesFile` for it. 

`livyConf` contains Livy options, and you should add app args like this; but you need two separate args, not a single arg concatenated with `+`.